analysis/tests/src/expected/TypeContextCompletion.res.txt

```diff
@@ -0,0 +1,194 @@
+Complete src/TypeContextCompletion.res 18:41
+posCursor:[18:41] posNoWhite:[18:40] Found expr:[18:11->18:41]
+Pexp_apply ...[18:11->18:30] (~someVaria18:32->18:41=...[18:32->18:41])
+Completable: CnamedArg(Value[someVariantToString], someVaria, [someVaria])
+Found type for function (~someVariant: someVariant) => string
+[{
+    "label": "someVariant",
+    "kind": 4,
+    "tags": [],
+    "detail": "someVariant",
+    "documentation": null
+  }]
+
+Complete src/TypeContextCompletion.res 21:44
+posCursor:[21:44] posNoWhite:[21:43] Found expr:[21:11->21:44]
+Pexp_apply ...[21:11->21:30] (~someVariant21:32->21:43=...__ghost__[0:-1->0:-1])
+found typed context 
+Completable: CtypedContext(Value[someVariantToString], NamedArg(someVariant))
+Found variant type for NamedArg typed context someVariant
+[{
+    "label": "One",
+    "kind": 4,
+    "tags": [],
+    "detail": "One\n\n",
+    "documentation": null
+  }, {
+    "label": "Two",
+    "kind": 4,
+    "tags": [],
+    "detail": "Two\n\n",
+    "documentation": null
+  }, {
+    "label": "Three",
+    "kind": 4,
+    "tags": [],
+    "detail": "Three\n\n",
+    "documentation": null
+  }, {
+    "label": "Four",
+    "kind": 4,
+    "tags": [],
+    "detail": "Four\n\n",
+    "documentation": null
+  }, {
+    "label": "Five(_)",
+    "kind": 4,
+    "tags": [],
+    "detail": "Five(int)\n\n",
+    "documentation": null
+  }, {
+    "label": "Six(_)",
+    "kind": 4,
+    "tags": [],
+    "detail": "Six(option<string>)\n\n",
+    "documentation": null
+  }]
+
+Complete src/TypeContextCompletion.res 24:45
+posCursor:[24:45] posNoWhite:[24:43] Found expr:[24:11->24:44]
+Pexp_apply ...[24:11->24:30] (~someVariant24:32->24:43=...__ghost__[0:-1->0:-1])
+found typed context 
+Completable: CtypedContext(Value[someVariantToString], NamedArg(someVariant))
+Found variant type for NamedArg typed context someVariant
+[{
+    "label": "One",
+    "kind": 4,
+    "tags": [],
+    "detail": "One\n\n",
+    "documentation": null
+  }, {
+    "label": "Two",
+    "kind": 4,
+    "tags": [],
+    "detail": "Two\n\n",
+    "documentation": null
+  }, {
+    "label": "Three",
+    "kind": 4,
+    "tags": [],
+    "detail": "Three\n\n",
+    "documentation": null
+  }, {
+    "label": "Four",
+    "kind": 4,
+    "tags": [],
+    "detail": "Four\n\n",
+    "documentation": null
+  }, {
+    "label": "Five(_)",
+    "kind": 4,
+    "tags": [],
+    "detail": "Five(int)\n\n",
```

Comment:
There's the question of what one would like to show here.
O I get it. Now I understand what "snippets" means. They're the templates for a variant with payloads?
If so then sure why not tackle them next.

Reply:
Yeah, essentially we could return this: `Five($1, $2)` and the editor would put the cursor on `$1` first, and then a tab would put the cursor on `$2`. This can be quite convenient when autocompleting things like this. For example, later when we do sub expression type aware completion (that's not a proper sentence...), the experience could be quite nice. Something like:
1. Complete `someFn(~somethingOptional=`, gives `Some($1)`
2. Select that completion, trigger autocomplete again. Get a list of all variant constructors if it's a variant, or all identifiers in scope matching the inner type of `somethingOptional`, etc

We do need to support both with and without snippets, because I'm unsure if all editors support them. I'll have a look at that later.

Reply:
Don't know if "sub expression type aware completion" is a proper sentence but is sure sounds cool.
What kind of format does vocode support for these kinds of templates? Does it expect things with dollar signs.
It would be great if there were a way to convey the type to, but don't know what format is possible and one would not like super-long names. But if the names were hoverable, then that would be pretty sweet.

Not sure if you want to go down the rabbit hole now, or move on with other aspects and come back to this later.

Reply:
We can come back later, there's other things snippet syntax would be nice to leverage for too. But yeah, too much of a side track now.